libr/util/spaces.c

```diff
@@ -1,213 +1,185 @@
+		goto fail;
 	}
-	for (i = 0; i < R_SPACES_MAX; i++) {
-		if (s->spaces[i] && !strcmp (name, s->spaces[i])) {
-			return i;
-		}
+
+	sp->event = r_event_new (sp);
+	if (!sp->event) {
+		goto fail;
 	}
-	return -1;
+
+	return true;
+
+fail:
+	r_spaces_free (sp);
+	return false;
 }
 
-R_API const char *r_space_get_i(RSpaces *s, int idx) {
-	if (idx==-1 || idx>=R_SPACES_MAX) {
-		return "";
-	}
-	return s->spaces[idx];
+R_API void r_spaces_free(RSpaces *sp) {
+	r_spaces_fini (sp);
+	free (sp);
 }
 
-R_API int r_space_add(RSpaces *s, const char *name) {
-	int i;
-	if (!name || *name == '*') {
-		return -1;
+static void space_free(RSpace *s) {
+	if (s) {
+		free (s->name);
+		free (s);
 	}
-	for (i = 0; i < R_SPACES_MAX; i++) {
-		if (s->spaces[i] && !strcmp (name, s->spaces[i])) {
-			return i;
-		}
-	}
-	// not found
-	for (i = 0; i < R_SPACES_MAX; i++) {
-		if (!s->spaces[i]) {
-			s->spaces[i] = strdup (name);
-			return i;
-		}
-	}
-	return -1;
 }
 
-R_API bool r_space_push(RSpaces *s, const char *name) {
-	int ret = false;
-	if (name && *name) {
-		if (s->space_idx >= 0 && s->spaces[s->space_idx]) {
-			r_list_push (s->spacestack, s->spaces[s->space_idx]);
-		} else {
-			r_list_push (s->spacestack, "*");
-		}
-		r_space_set (s, name);
-		ret = true;
-	}
-	return ret;
+static void space_node_free(RBNode *n) {
+	RSpace *s = container_of (n, RSpace, rb);
+	space_free (s);
 }
 
-R_API bool r_space_pop(RSpaces *s) {
-	char *p = r_list_pop (s->spacestack);
-	if (p) {
-		if (*p) {
-			r_space_set (s, p);
-		}
-		return true;
-	}
-	return false;
+R_API void r_spaces_fini(RSpaces *sp) {
+	r_list_free (sp->spacestack);
+	sp->spacestack = NULL;
+	r_rbtree_free (sp->spaces, space_node_free);
+	sp->spaces = NULL;
+	r_event_free (sp->event);
+	sp->event = NULL;
+	R_FREE (sp->name);
 }
 
-R_API int r_space_set(RSpaces *s, const char *name) {
-	s->space_idx = r_space_add (s, name);
-	return s->space_idx;
+static int name_space_cmp(const void *incoming, const RBNode *rb) {
+	const RSpace *s = container_of (rb, const RSpace, rb);
+	return strcmp (incoming, s->name);
 }
 
-R_API int r_space_unset (RSpaces *s, const char *name) {
-	int i, count = 0;
-	if (!name) {
-		r_space_set (s, NULL);
-	}
-	for (i = 0; i < R_SPACES_MAX; i++) {
-		if (!s->spaces[i]) {
-			continue;
-		}
-		if (!name || !strcmp (name, s->spaces[i])) {
-			if (s->space_idx == i) {
-				s->space_idx = -1;
-			}
-			R_FREE (s->spaces[i]);
-			if (s->unset_for) {
-				s->unset_for (s, i);
-			}
-			count++;
-		}
-	}
-	return count;
+R_API RSpace *r_spaces_get(RSpaces *sp, const char *name) {
+	RBNode *n = r_rbtree_find (sp->spaces, (void *)name, name_space_cmp);
+	return n? container_of (n, RSpace, rb): NULL;
 }
 
-static int r_space_count (RSpaces *s, int n) {
-	if (s->count_for) {
-		return s->count_for (s, n);
-	}
-	return 0;
+static int space_cmp(const void *incoming, const RBNode *rb) {
+	const RSpace *a = (const RSpace *)incoming;
+	const RSpace *b = container_of (rb, const RSpace, rb);
```

Comment:
though this container_of would be renamed before using it in more places

Reply:
we agreed it was enough to move it to r_util.h. container_of is a very generic function, used also in the kernel. I really think the name is perfect. We can create an additional macro specific for RSpace (that uses container_of) if we want.